File: backend/app/schemas/student.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from app.models.student_data import ComplaintCategory, ComplaintStatus


class CitationSource(BaseModel):
    document_title: str
    section: Optional[str] = None
    chunk_text: Optional[str] = None
    relevance_score: float = 0.0
    url: Optional[str] = None


class StudentBase(BaseModel):
    program: str
    semester: int
    section: str
    batch_year: int
    cgpa: Optional[float] = None


class StudentProfileResponse(StudentBase):
    id: int
    user_id: int
    full_name: str
    email: str
    university_id: str

    class Config:
        from_attributes = True


class ComplaintCreate(BaseModel):
    category: ComplaintCategory
    title: str = Field(..., min_length=5, max_length=200)
    description: str = Field(..., min_length=10)


class ComplaintStatusUpdate(BaseModel):
    status: ComplaintStatus
    assigned_to: Optional[str] = None
    resolution_notes: Optional[str] = None


class ComplaintSchema(BaseModel):
    id: int
    ticket_id: str
    category: ComplaintCategory
    title: str
    description: str
    status: ComplaintStatus
    assigned_to: Optional[str] = None
    resolution_notes: Optional[str] = None
    created_at: datetime
    resolved_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ChatMessage(BaseModel):
    role: str
    content: str
    sources: Optional[List[CitationSource]] = None
    data_sources: Optional[List[str]] = None
    timestamp: Optional[str] = None


class ChatRequest(BaseModel):
    message: str
    conversation_id: Optional[str] = None
    history: Optional[List[ChatMessage]] = []


class ChatResponse(BaseModel):
    message: str
    sources: List[CitationSource] = []
    data_sources: List[str] = []
    conversation_id: str


class AdminStatsSchema(BaseModel):
    total_students: int
    total_complaints: int
    open_complaints: int
    resolved_complaints: int

File: frontend/src/types/student.ts
export type ComplaintCategory = 'HOSTEL' | 'ACADEMIC' | 'INFRASTRUCTURE' | 'ADMINISTRATIVE' | 'OTHER';
export type ComplaintStatus = 'SUBMITTED' | 'IN_REVIEW' | 'RESOLVED' | 'CLOSED';

export interface StudentProfile {
  id: number;
  user_id: number;
  full_name: string;
  email: string;
  university_id: string;
  program: string;
  semester: number;
  section: string;
  batch_year: number;
  cgpa?: number;
}

export interface Complaint {
  id: number;
  ticket_id: string;
  category: ComplaintCategory;
  title: string;
  description: string;
  status: ComplaintStatus;
  assigned_to?: string;
  resolution_notes?: string;
  created_at: string;
  resolved_at?: string;
}

export interface CreateComplaintInput {
  category: ComplaintCategory;
  title: string;
  description: string;
}

export interface SourceCitation {
  document_title?: string;
  title?: string;
  section?: string;
  chunk_text?: string;
  snippet?: string;
  relevance_score: number;
  url?: string;
}

export type ChatSource = SourceCitation;

export interface ChatMessage {
  id?: string;
  role: 'user' | 'assistant';
  content: string;
  sources?: SourceCitation[];
  data_sources?: string[];
  timestamp?: string;
}

export interface ChatResponse {
  message?: string;
  answer?: string;
  sources: SourceCitation[];
  data_sources?: string[];
  conversation_id: string;
}

export interface AdminStats {
  total_students: number;
  total_faculty?: number;
  total_users?: number;
  total_documents?: number;
  total_complaints?: number;
  open_complaints?: number;
  resolved_complaints?: number;
}

export interface UserItem {
  id: number;
  email: string;
  full_name: string;
  role: string;
  university_id: string;
  is_active: boolean;
  created_at?: string;
}

export interface DocumentItem {
  id: number;
  title: string;
  category: string;
  blob_url: string;
  uploaded_at: string;
}
